src/datasources/notifications/notifications.datasource.spec.ts

```diff
@@ -0,0 +1,633 @@
+
+      const actual = await target.upsertSubscriptions({
+        signerAddress,
+        upsertSubscriptionsDto,
+      });
+
+      expect(actual).toStrictEqual({ deviceUuid: expect.any(String) });
+
+      // Ensure correct database structure
+      await Promise.all([
+        sql`SELECT * FROM push_notification_devices`,
+        sql`SELECT * FROM notification_types`,
+        sql`SELECT * FROM notification_subscriptions`,
+        sql`SELECT * FROM notification_subscription_notification_types`,
+      ]).then(([devices, types, subscriptions, subscribedNotifications]) => {
+        expect(devices).toStrictEqual([
+          {
+            id: 1,
+            device_type: upsertSubscriptionsDto.deviceType,
+            device_uuid: actual.deviceUuid,
+            cloud_messaging_token: upsertSubscriptionsDto.cloudMessagingToken,
+            created_at: expect.any(Date),
+            updated_at: expect.any(Date),
+          },
+        ]);
+        expect(types).toStrictEqual(
+          Object.values(NotificationType).map((type) => {
+            return {
+              id: expect.any(Number),
+              name: type,
+            };
+          }),
+        );
+        expect(subscriptions).toStrictEqual(
+          upsertSubscriptionsDto.safes.map((safe, i) => {
+            return {
+              id: i + 1,
+              signer_address: signerAddress,
+              push_notification_device_id: devices[0].id,
+              chain_id: safe.chainId,
+              safe_address: safe.address,
+              created_at: expect.any(Date),
+              updated_at: expect.any(Date),
+            };
+          }),
+        );
+        expect(subscribedNotifications).toStrictEqual(
+          expect.arrayContaining(
+            upsertSubscriptionsDto.safes.flatMap((safe, i) => {
+              return safe.notificationTypes.map((type) => {
+                return {
+                  id: expect.any(Number),
+                  notification_subscription_id: i + 1,
+                  notification_type_id: types.find((t) => t.name === type)?.id,
+                };
+              });
+            }),
+          ),
+        );
+      });
+    });
+
+    it('should always update the deviceType/cloudMessagingToken', async () => {
+      const signerAddress = getAddress(faker.finance.ethereumAddress());
+      const upsertSubscriptionsDto = upsertSubscriptionsDtoBuilder().build();
+      const secondSubscriptionsDto = upsertSubscriptionsDtoBuilder()
+        .with('deviceUuid', upsertSubscriptionsDto.deviceUuid)
+        .build();
+      await target.upsertSubscriptions({
+        signerAddress,
+        upsertSubscriptionsDto,
+      });
+
+      // Insert should not throw despite it being the same device UUID
+      await expect(
+        target.upsertSubscriptions({
+          signerAddress,
+          upsertSubscriptionsDto: secondSubscriptionsDto,
+        }),
+      ).resolves.not.toThrow();
+      // Device UUID should have updated
```

Comment:
Do you mean `shouldn't have been updated`?

Reply:
Fixed in d49aca2e6a097d8ac438fa5876d39f1b2006bf3f.